### PR TITLE
Add combination fv_converter

### PR DESCRIPTION
**Pull request by hido**
This PR fixes [#61](https://github.com/jubatus/jubatus/issues/61).
Config JSON now has ``combination_types`` and ``combination_rules``.
For each key pairs specified by ``key_left`` and ``key_right``, combinational feature value will be computed.
Note that ``key_left`` and ``key_right`` looks at the keys of ``sfv_t``, not those of the original datum.
This process is executed after num/string/binary rules and before feature hashing.

### CODE REVIEW

**Comment by hido** [comment]
The config and example can be found in [jubatus-example/shogun](https://github.com/hido/jubatus-example/blob/12b4c7e09dfdd6deecbd263f7bd42c125b6e214f/shogun/shogun.json)

File: jubatus/core/fv_converter/combination_feature.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include <utility>
+#include <vector>
+
+namespace jubatus {
+namespace core {
+namespace fv_converter {
+
+class combination_feature {
+ public:
+  combination_feature() {
```
Comment (gwtnb) [inline jubatus/core/fv_converter/combination_feature.hpp]: This default constructor seems verbose.

File: jubatus/core/fv_converter/datum_to_fv_converter_test.cpp [hunk 2]
```diff
@@ -442,6 +445,102 @@ TEST(datum_to_fv_converter, check_datum_key_in_binary) {
   ASSERT_THROW(conv.convert(datum, feature), converter_exception);
 }
 
+TEST(datum_to_fv_converter, combination_feature_num) {
+  datum datum;
+  datum.num_values_.push_back(std::make_pair("/val1", 1.1));
```
Comment (kumagi) [inline jubatus/core/fv_converter/datum_to_fv_converter_test.cpp]: why `/` prefix? some comment may be needed.

File: jubatus/core/fv_converter/datum_to_fv_converter_test.cpp [hunk 3]
```diff
@@ -442,6 +445,102 @@ TEST(datum_to_fv_converter, check_datum_key_in_binary) {
...
+      "mul",
+      all_matcher,
+      all_matcher,
+      combination_feature_t(new combination_mul_feature()));
+  std::vector<std::pair<std::string, float> > feature;
+  conv.convert(datum, feature);
+
+  std::vector<std::pair<std::string, float> > expected;
+  expected.push_back(std::make_pair("/val1@num", 1.1));
+  expected.push_back(std::make_pair("/val2@num", 1.1));
```
Comment (kumagi) [inline jubatus/core/fv_converter/datum_to_fv_converter_test.cpp]: How about using different number from `val1`. For confirmation that it is not using `val1` twice.
This test will be wrongly passed although we implementd 
```
void add_feature(const std::string& key,
  double value_left,
  double value_right,
  common::sfv_t& ret_fv) const {
  ret_fv.push_back(
   std::make_pair(key, static_cast<float>(value_left * value_left)));  // using left twice!!!!
}
```

File: jubatus/core/fv_converter/combination_feature_factory.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    combination_feature*(const std::string&, const param_t&)> create_function;
+
+  combination_feature_factory() {
+  }
+
+  explicit combination_feature_factory(const create_function& ext)
+    : ext_(ext) {
+  }
+
+  combination_feature* create(const std::string& name,
```
Comment (gwtnb) [inline jubatus/core/fv_converter/combination_feature_factory.hpp]: What is the reason to use a raw pointer here? num_filter_factory uses shared_ptr. https://github.com/jubatus/jubatus_core/blob/master/jubatus/core/fv_converter/num_filter_factory.hpp#L44

File: jubatus/core/fv_converter/datum_to_fv_converter.cpp [hunk 5]
```diff
@@ -501,6 +539,27 @@ class datum_to_fv_converter_impl {
...
+  void convert_combinations(common::sfv_t& ret_fv) const {
+    size_t original_size = ret_fv.size();
+    for (size_t i = 0; i < combination_rules_.size(); ++i) {
+      const combination_feature_rule& r = combination_rules_[i];
+      for (size_t j = 0 ; j < original_size - 1; ++j) {
+        for (size_t m = j + 1; m < original_size; ++m) {
+          if (r.matcher_left_->match(ret_fv[j].first)
+              && r.matcher_right_->match(ret_fv[m].first)) {
+            std::string k =
+                ret_fv[j].first + "&" + ret_fv[m].first + "/" + r.name_;
```
Comment (kumagi) [inline jubatus/core/fv_converter/datum_to_fv_converter.cpp]: This variable is declared but used just one time below.
And depth of indentation is the same. How about merge?

File: jubatus/core/fv_converter/datum_to_fv_converter.cpp [hunk 6]
```diff
@@ -501,6 +539,27 @@ class datum_to_fv_converter_impl {
       }
     }
   }
+
+  void convert_combinations(common::sfv_t& ret_fv) const {
+    size_t original_size = ret_fv.size();
```
Comment (kumagi) [inline jubatus/core/fv_converter/datum_to_fv_converter.cpp]: const.

**Comment by kumagi** [comment]
By the way, `weight_manager` see naiiively all features with terminated with `idf` keyword.
How should we treat combination of feature with `idf` weighted?

File: jubatus/core/fv_converter/combination_feature.hpp (hunk 1 above)
Comment (hido) [inline jubatus/core/fv_converter/combination_feature.hpp]: Removed.

File: jubatus/core/fv_converter/combination_feature_factory.hpp (hunk 4 above)
Comment (hido) [inline jubatus/core/fv_converter/combination_feature_factory.hpp]: Use shared_ptr instead.

File: jubatus/core/fv_converter/datum_to_fv_converter.cpp (hunk 6 above)
Comment (hido) [inline jubatus/core/fv_converter/datum_to_fv_converter.cpp]: Added.

File: jubatus/core/fv_converter/datum_to_fv_converter.cpp (hunk 5 above)
Comment (hido) [inline jubatus/core/fv_converter/datum_to_fv_converter.cpp]: Fixed.

File: jubatus/core/fv_converter/datum_to_fv_converter_test.cpp (hunk 3 above)
Comment (hido) [inline jubatus/core/fv_converter/datum_to_fv_converter_test.cpp]: Fixed with different values.

**Comment by kumagi** [comment]
↑ it is needles fear. `convert_combination()` function is called just after `weight_manger` worked.
`weight_manager` can't see combination features.

**Comment by kumagi** [comment]
note: need documentation.

**Comment by kumagi** [comment]
LGTM!

**Comment by kmaehashi** [comment]
> note: need documentation.

Raised issue for that.
https://github.com/jubatus/website/issues/174

**Pull request merged**